# Trabajo Práctico 3

In [2]:
%%writefile README.md

ALLENDE, Lucila Abigail
SERRUYA ALOISI, Luciano Sebastián
TOLEDO MARGALEF, Pablo Adrián

Overwriting README.md


## Ejercicio 1: multiplicación de matrices cuadradas (secuencial)

### Orden de las matrices: 2000

In [7]:
%%writefile mmCompleta.c

#include <stdio.h>

#define N 2000

double a[N][N];
double b[N][N];
double c[N][N];

void iniMat(double mat[][N], int len, double iniVal)
{
	int i, j;

	for(i = 0; i < len; ++i)
		for(j = 0; j < len; ++j)
			mat[i][j] = iniVal;
}

/* Check1Mat: Chequea que la matriz resultante quede con
 *            el mismo valor en todas sus celdas.
 * Es un chequeo básico para el caso particular en que
 * se sabe de ante mano que el resultado tendrá esa forma.
 * Retorna: 0 en caso correcto (todas las celdas contiene
 *          finVal. 1 En caso de error (al menos una celda
 * contiene un valor diferente de finVal).
 */
int Check1Mat(double mat[][N], int len, double finVal)
{
	int i, j;

	for(i = 0; i < len; ++i)
		for(j = 0; j < len; ++j)
			if(mat[i][j] != finVal)
				return 1;
	return 0;
}

void printMat(double mat[][N], int len)
{
	int i, j;

	for(i = 0; i < len; ++i)
	{
		for(j = 0; j < len; ++j)
			printf( "%.1f ", mat[i][j]);
		printf("\n");
	}
}

void mm(void)
{
	int i, j, k;

	for(i = 0; i < N; ++i)
		for(j = 0; j < N; ++j)
			for(k = 0; k < N; ++k)
				c[i][j] += a[i][k] * b[k][j];
}

int main(void)
{
	printf("Comienzo ...\n");
	iniMat(a, N, 1.0);
	iniMat(b, N, 1.0);
	iniMat(c, N, 0.0);

	printf("Multiplicando ...\n");

	mm();

	if(Check1Mat(c, N, N) == 0)
		printf("Fin Multiplicación (Resultado correcto)\n");
	else
		printf("Fin Multiplicación (Resultado INCORRECTO!)\n");

/*
	printf("Matriz a:\n");
	printMat(a, N);
	printf("Matriz b:\n");
	printMat(b, N);
	printf("Matriz c:\n");
	printMat(c, N);
*/

	printf("Fin del programa\n");
}


Writing mmCompleta.c


In [8]:
!gcc mmCompleta.c -o mmCompleta

#### Pasamos a realizar las pruebas de tiempo. Corriendo tres veces seguidas el programa sin carga del sistema, tres veces y haciendo un promedio de los tres tiempos

In [5]:
!time ./mmCompleta && time ./mmCompleta && time ./mmCompleta

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.19user 0.03system 1:06.24elapsed 99%CPU (0avgtext+0avgdata 94956maxresident)k
0inputs+0outputs (0major+1014minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.04user 0.08system 1:06.13elapsed 99%CPU (0avgtext+0avgdata 95016maxresident)k
0inputs+0outputs (0major+1015minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.11user 0.07system 1:06.21elapsed 99%CPU (0avgtext+0avgdata 94968maxresident)k
0inputs+0outputs (0major+1015minor)pagefaults 0swaps


`Promedio de tiempo: (66.24 + 66.13 + 66.2) / 3 = 66.19 = 1:06.19` 


#### Ahora realizamos la misma prueba con una matriz de la mitad del tamaño de la anterior

In [23]:
%%writefile mmMitad.c

#include <stdio.h>

#define N 1000

double a[N][N];
double b[N][N];
double c[N][N];

void iniMat(double mat[][N], int len, double iniVal)
{
	int i, j;

	for(i = 0; i < len; ++i)
		for(j = 0; j < len; ++j)
			mat[i][j] = iniVal;
}

/* Check1Mat: Chequea que la matriz resultante quede con
 *            el mismo valor en todas sus celdas.
 * Es un chequeo básico para el caso particular en que
 * se sabe de ante mano que el resultado tendrá esa forma.
 * Retorna: 0 en caso correcto (todas las celdas contiene
 *          finVal. 1 En caso de error (al menos una celda
 * contiene un valor diferente de finVal).
 */
int Check1Mat(double mat[][N], int len, double finVal)
{
	int i, j;

	for(i = 0; i < len; ++i)
		for(j = 0; j < len; ++j)
			if(mat[i][j] != finVal)
				return 1;
	return 0;
}

void printMat(double mat[][N], int len)
{
	int i, j;

	for(i = 0; i < len; ++i)
	{
		for(j = 0; j < len; ++j)
			printf( "%.1f ", mat[i][j]);
		printf("\n");
	}
}

void mm(void)
{
	int i, j, k;

	for(i = 0; i < N; ++i)
		for(j = 0; j < N; ++j)
			for(k = 0; k < N; ++k)
				c[i][j] += a[i][k] * b[k][j];
}

int main(void)
{
	printf("Comienzo ...\n");
	iniMat(a, N, 1.0);
	iniMat(b, N, 1.0);
	iniMat(c, N, 0.0);

	printf("Multiplicando ...\n");

	mm();

	if(Check1Mat(c, N, N) == 0)
		printf("Fin Multiplicación (Resultado correcto)\n");
	else
		printf("Fin Multiplicación (Resultado INCORRECTO!)\n");

/*
	printf("Matriz a:\n");
	printMat(a, N);
	printf("Matriz b:\n");
	printMat(b, N);
	printf("Matriz c:\n");
	printMat(c, N);
*/

	printf("Fin del programa\n");
}


Overwriting mmMitad.c


In [24]:
!gcc mmMitad.c -o mmMitad

In [25]:
!time ./mmMitad && time ./mmMitad && time ./mmMitad

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
7.77user 0.01system 0:07.81elapsed 99%CPU (0avgtext+0avgdata 24656maxresident)k
0inputs+0outputs (0major+811minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
7.80user 0.00system 0:07.81elapsed 99%CPU (0avgtext+0avgdata 24656maxresident)k
0inputs+0outputs (0major+813minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
7.78user 0.01system 0:07.80elapsed 99%CPU (0avgtext+0avgdata 24740maxresident)k
0inputs+0outputs (0major+812minor)pagefaults 0swaps


`Promedio de ejecución: (7.81 + 7.81 + 7.80) / 3 = 7.806`

### Ejercicio 2: Pruebas con banderas

In [36]:
!gcc mmCompleta.c -o mmCompleta -O0

In [37]:
!time ./mmCompleta && time ./mmCompleta && time ./mmCompleta

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.47user 0.02system 1:06.54elapsed 99%CPU (0avgtext+0avgdata 95012maxresident)k
0inputs+0outputs (0major+1016minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.13user 0.01system 1:06.16elapsed 99%CPU (0avgtext+0avgdata 95008maxresident)k
0inputs+0outputs (0major+1014minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.31user 0.04system 1:06.37elapsed 99%CPU (0avgtext+0avgdata 95124maxresident)k
0inputs+0outputs (0major+1017minor)pagefaults 0swaps


In [38]:
!gcc mmCompleta.c -o mmCompleta -O1

In [39]:
!time ./mmCompleta && time ./mmCompleta && time ./mmCompleta

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
22.26user 0.04system 0:22.31elapsed 99%CPU (0avgtext+0avgdata 94916maxresident)k
0inputs+0outputs (0major+1015minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
22.24user 0.00system 0:22.27elapsed 99%CPU (0avgtext+0avgdata 94984maxresident)k
0inputs+0outputs (0major+1016minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
22.23user 0.02system 0:22.26elapsed 99%CPU (0avgtext+0avgdata 95008maxresident)k
0inputs+0outputs (0major+1013minor)pagefaults 0swaps


In [40]:
!gcc mmCompleta.c -o mmCompleta -O2

In [41]:
!time ./mmCompleta && time ./mmCompleta && time ./mmCompleta

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
22.65user 0.02system 0:22.70elapsed 99%CPU (0avgtext+0avgdata 95000maxresident)k
0inputs+0outputs (0major+1015minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
22.69user 0.02system 0:22.73elapsed 99%CPU (0avgtext+0avgdata 95056maxresident)k
0inputs+0outputs (0major+1018minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
22.78user 0.02system 0:22.84elapsed 99%CPU (0avgtext+0avgdata 95052maxresident)k
0inputs+0outputs (0major+1016minor)pagefaults 0swaps


In [42]:
!gcc mmCompleta.c -o mmCompleta -O3

In [43]:
!time ./mmCompleta && time ./mmCompleta && time ./mmCompleta

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
15.63user 0.02system 0:15.69elapsed 99%CPU (0avgtext+0avgdata 94980maxresident)k
0inputs+0outputs (0major+1016minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
15.64user 0.02system 0:15.67elapsed 99%CPU (0avgtext+0avgdata 94988maxresident)k
0inputs+0outputs (0major+1014minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
15.76user 0.02system 0:15.79elapsed 99%CPU (0avgtext+0avgdata 95024maxresident)k
0inputs+0outputs (0major+1016minor)pagefaults 0swaps


### Ejercicio 3: Matriz completa en paralelo

In [44]:
%%writefile ej3.c

/* Introducción a la Concurrencia 2016
 *
 * Ejemplo de multiplicación de matrices.
 * Programa secuencial.
 *
 */

#include <stdio.h>
#include <omp.h>

#define N 2000

double a[N][N];
double b[N][N];
double c[N][N];

void iniMat(double mat[][N], int len, double iniVal)
{
	for(int i = 0; i < len; ++i)
		for(int j = 0; j < len; ++j)
			mat[i][j] = iniVal;
}

/* Check1Mat: Chequea que la matriz resultante quede con
 *            el mismo valor en todas sus celdas.
 * Es un chequeo básico para el caso particular en que
 * se sabe de ante mano que el resultado tendrá esa forma.
 * Retorna: 0 en caso correcto (todas las celdas contiene
 *          finVal. 1 En caso de error (al menos una celda
 * contiene un valor diferente de finVal).
 */
int Check1Mat(double mat[][N], int len, double finVal)
{
	int i, j;

	for(i = 0; i < len; ++i)
		for(j = 0; j < len; ++j)
			if(mat[i][j] != finVal)
				return 1;
	return 0;
}
void printMat(double mat[][N], int len)
{
	int i, j;

	for(i = 0; i < len; ++i)
	{
		for(j = 0; j < len; ++j)
			printf( "%.1f ", mat[i][j]);
		printf("\n");
	}
}

void mm(void){        
    
    #pragma omp parallel for
    //maneja las filas
    for(int i = 0; i < N; ++i){
        //maneja las columnas
        for(int j = 0; j < N; ++j){
            //calcular una celda de la resultante
            for(int k = 0; k < N; ++k){
                c[i][j] += a[i][k] * b[k][j];   
            }
        }             
    }

}

int main(void)
{
	printf("Comienzo ...\n");
	iniMat(a, N, 1.0);
	iniMat(b, N, 1.0);
	iniMat(c, N, 0.0);

	printf("Multiplicando ...\n");

	mm();

	if(Check1Mat(c, N, N) == 0)
		printf("Fin Multiplicación (Resultado correcto)\n");
	else
		printf("Fin Multiplicación (Resultado INCORRECTO!)\n");

	printf("Fin del programa\n");
}


Writing ej3.c


In [47]:
!gcc ej3.c -o ej3 -fopenmp

In [49]:
!time ./ej3 && time ./ej3 && time ./ej3

Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.36user 0.10system 0:34.72elapsed 191%CPU (0avgtext+0avgdata 95400maxresident)k
0inputs+0outputs (0major+1549minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.44user 0.05system 0:34.66elapsed 191%CPU (0avgtext+0avgdata 95260maxresident)k
0inputs+0outputs (0major+1035minor)pagefaults 0swaps
Comienzo ...
Multiplicando ...
Fin Multiplicación (Resultado correcto)
Fin del programa
66.54user 0.04system 0:34.76elapsed 191%CPU (0avgtext+0avgdata 95344maxresident)k
0inputs+0outputs (0major+1040minor)pagefaults 0swaps


`Tiempo promedio: 34.71 seg`

#### La paralelización de la matriz se realiza por filas.  
#### Se aprovecharía al máximo la propiedad de la localidad de referencia, si se disponen los elementos de la matriz A por filas, y los elementos de la matriz B por columnas, es decir, su traspuesta.

### Ejercicio 4: Cree un programa que muestre el ID de cada hilo en un bloque paralelo de OpenMP

In [73]:
%%writefile ej4.c 

#include <omp.h>
#include <stdio.h>

int main(void){
    
    #pragma omp parallel 
    {
        int ID = omp_get_thread_num();
        printf("Hilo:%d\n", ID);
    }
}

Overwriting ej4.c


In [74]:
!gcc ej4.c -o ej4 -fopenmp

In [75]:
!./ej4


Hilo:0
Hilo:1


### Ejercicio 5: Sumatoria de arreglo (1000 posiciones)

In [98]:
%%writefile ej5.c

#include <stdio.h>
#include <unistd.h>
#define MAX 1000


void iniArreglo(int *a){
    int i;
    for(i = 0; i < MAX; i++){
        a[i] = i;
    }
}

int sumatoria(int *a){
    int valor = 0;
    int i;
    
    for(i = 0; i < MAX; i++){
        valor += a[i];
        usleep(10);
        /*agregamos espera para que se note la diferencia
        entre la ejecución paralela y secuencial*/
    }
    return valor;
}

int main(void){
    int arreglo[MAX];
    
    iniArreglo(arreglo);
    
    int acum = sumatoria(arreglo);
    
    printf("Valor del acumulador: %d\n", acum);
    
    return 0;
}

Overwriting ej5.c


In [99]:
!gcc ej5.c -o ej5

In [100]:
!time ./ej5

Valor del acumulador: 499500
0.00user 0.00system 0:00.06elapsed 0%CPU (0avgtext+0avgdata 1380maxresident)k
0inputs+0outputs (0major+65minor)pagefaults 0swaps


### Ejercicio 6: Sumatoria paralela

In [101]:
%%writefile ej6.c

#include <stdio.h>
#include <omp.h>
#include <unistd.h>

#define MAX 1000


void iniArreglo(int *a){
    int i;
    for(i = 0; i < MAX; i++){
        a[i] = i;
    }
}

int sumatoria(int *a){
    int valor;
    int i;
    
    #pragma omp parallel for reduction(+:valor)
    for(i = 0; i < MAX; i++){
        valor += a[i];
        usleep(10);  
        /*agregamos espera para que se note la diferencia
        entre la ejecución paralela y secuencial*/
    }
    return valor;
}

int main(void){
    int arreglo[MAX];
    
    iniArreglo(arreglo);
    
    int acum = sumatoria(arreglo);
    
    printf("Valor del acumulador: %d\n", acum);
    
    return 0;
}

Overwriting ej6.c


In [102]:
!gcc ej6.c -o ej6 -fopenmp

In [103]:
!time ./ej6

Valor del acumulador: 499500
0.00user 0.00system 0:00.03elapsed 21%CPU (0avgtext+0avgdata 1720maxresident)k
0inputs+0outputs (0major+90minor)pagefaults 0swaps


### Ejercicio 7: Implementación (secuencial) de integral

In [4]:
%%writefile ej7.c

#include<stdio.h>
#include<unistd.h>

int num_steps = 100000;
double step;
int main(){
    int i;
    double x, pi, sum = 0.0;
    double incremento = (double)num_steps / (num_steps * 2); 
    
    step = 1.0/(double)num_steps;

    
    for(i=0; i < num_steps; i++){
        x = (i + incremento)*step;
        sum += 4.0/(1.0+(x*x));
        usleep(1);
    }
    
    pi = step*sum;
    
    printf("PI = %.50f\n", pi);
    return 0;
}

Overwriting ej7.c


In [5]:
!gcc ej7.c -o ej7

In [6]:
!time ./ej7

PI = 3.14159265359816153306837804848328232765197753906250
0.03user 0.32system 0:05.71elapsed 6%CPU (0avgtext+0avgdata 1372maxresident)k
0inputs+0outputs (0major+63minor)pagefaults 0swaps


### Ejercicio 8: Implementación (paralela) de integral

In [1]:
%%writefile ej8.c

#include<stdio.h>
#include<unistd.h>
#include<omp.h>

int num_steps = 100000;
double step;
int main(){
    int i;
    double x, pi, sum;
    double incremento = (double)num_steps / (num_steps * 2); 
    
    step = 1.0/(double)num_steps;

    #pragma omp parallel for reduction(+:sum)
    for(i=0; i < num_steps; i++){
        x = (i + incremento)*step;
        sum += 4.0/(1.0+(x*x));        
        usleep(1);
    }
    
    pi = step*sum;
    
    printf("PI = %.50f\n", pi);
    return 0;
}

Overwriting ej8.c


In [2]:
!gcc ej8.c -o ej8 -fopenmp

In [3]:
!time ./ej8

PI = 3.14159265359812556184238019341137260198593139648438
0.04user 0.11system 0:00.76elapsed 20%CPU (0avgtext+0avgdata 1772maxresident)k
0inputs+0outputs (0major+96minor)pagefaults 0swaps


### Ejercicio 9: investigación sobre compiladores y lenguajes que soportan OpenMP

    Proveedor: Oracle
    Compilador: C/C++ / Fortran
    Bandera de compilación correspondiente a OMP: -xopenmp (Solaris y Linux)
    
    Proveedor: Intel
    Compilador: C/C++ / Fortran
    Bandera de compilación correspondiente a OMP: -Qopenmp (Windows); -openmp (Linux y Mac OSX)
    
    Proveedor: PGI
    Compilador: C/C++ / Fortran
    Bandera de compilación correspondiente a OMP: -mp (todas las plataformas)
    
    
    OpenMP está diseñado para Fortran, C y C++. Otros lenguajes implementan ciertas herramientas y extensiones (no oficiales de OpenMP) para utilizar las bibliotecas y directivas de OMP. 
    
    Algunos ejemplos son:
    
    Cython: lenguaje de programación (superconjunto de Python) que une Python con el sistema de tipado estático de C y C++.
    
    omp4j: preprocesador de código abierto similiar a OpenMP para Java 